File: aimDK/src/examples/src/hal/echo_imu_data.cpp
#include <chrono>
#include <iomanip>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <string>

class ImuPrinter : public rclcpp::Node {
public:
  ImuPrinter() : rclcpp::Node("imu_printer") {
    // Allow overriding the topic name via a parameter
    imu_topic_ = this->declare_parameter<std::string>(
        "imu_topic",
        "/aima/hal/imu/chest/state"); // Change to other IMU topics, e.g.
                                      // waist IMU: /aima/hal/imu/torse/state
                                      // lidar IMU:
                                      // /aima/hal/sensor/lidar_chest_front/imu

    auto qos = rclcpp::SensorDataQoS(); // BestEffort / Volatile
    sub_ = this->create_subscription<sensor_msgs::msg::Imu>(
        imu_topic_, qos,
        std::bind(&ImuPrinter::cb, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "Subscribing IMU: %s", imu_topic_.c_str());
  }

private:
  void cb(const sensor_msgs::msg::Imu::SharedPtr msg) {
    // Compute receive interval (based on wall time, for observation only)
    auto now = std::chrono::steady_clock::now();
    double dt_ms =
        last_recv_.time_since_epoch().count() == 0
            ? 0.0
            : std::chrono::duration<double, std::milli>(now - last_recv_)
                  .count();
    last_recv_ = now;

    const auto &q = msg->orientation;
    const auto &w = msg->angular_velocity;
    const auto &a = msg->linear_acceleration;

    // Print
    std::ostringstream oss;
    oss << std::fixed << std::setprecision(6);
    oss << "stamp=" << rclcpp::Time(msg->header.stamp).seconds()
        << "s  frame=" << msg->header.frame_id
        << "  recv_dt=" << std::setprecision(3) << dt_ms << "ms\n"
        << std::setprecision(6) << "  orientation (x,y,z,w): [" << q.x << ", "
        << q.y << ", " << q.z << ", " << q.w << "]\n"
        << "  angular_velocity (rad/s): [" << w.x << ", " << w.y << ", " << w.z
        << "]\n"
        << "  linear_accel (m/s^2): [" << a.x << ", " << a.y << ", " << a.z
        << "]";

    // Optional: print diagonal covariances (-1 means unknown)
    auto oc0 = msg->orientation_covariance[0];
    auto oc4 = msg->orientation_covariance[4];
    auto oc8 = msg->orientation_covariance[8];
    auto wc0 = msg->angular_velocity_covariance[0];
    auto wc4 = msg->angular_velocity_covariance[4];
    auto wc8 = msg->angular_velocity_covariance[8];
    auto ac0 = msg->linear_acceleration_covariance[0];
    auto ac4 = msg->linear_acceleration_covariance[4];
    auto ac8 = msg->linear_acceleration_covariance[8];

    oss << "\n  cov_diag(ori, ang, acc): [" << oc0 << "," << oc4 << "," << oc8
        << " | " << wc0 << "," << wc4 << "," << wc8 << " | " << ac0 << ","
        << ac4 << "," << ac8 << "]";

    RCLCPP_INFO(this->get_logger(), "%s", oss.str().c_str());
  }

  std::string imu_topic_;
  rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr sub_;
  std::chrono::steady_clock::time_point last_recv_{};
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ImuPrinter>());
  rclcpp::shutdown();
  return 0;
}

File: aimDK/src/py_examples/py_examples/echo_imu_data.py
#!/usr/bin/env python3
import time
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from rclpy.time import Time
from sensor_msgs.msg import Imu


class ImuPrinterPy(Node):
    def __init__(self):
        super().__init__('imu_printer')
        self.declare_parameter('imu_topic', '/aima/hal/imu/chest/state')
        # Can be replaced with other IMU topic names.
        # Waist IMU: /aima/hal/imu/torse/state
        # LiDAR IMU: /aima/hal/sensor/lidar_chest_front/imu
        self.topic = self.get_parameter(
            'imu_topic').get_parameter_value().string_value

        self._last_recv = None
        self.sub = self.create_subscription(
            Imu, self.topic, self.cb, qos_profile_sensor_data
        )
        self.get_logger().info(f'Subscribing IMU: {self.topic}')

    def cb(self, msg: Imu):
        now = time.perf_counter()
        dt_ms = 0.0 if self._last_recv is None else (
            now - self._last_recv) * 1000.0
        self._last_recv = now

        t_sec = Time.from_msg(msg.header.stamp).nanoseconds / 1e9
        q = msg.orientation
        w = msg.angular_velocity
        a = msg.linear_acceleration
        oc = msg.orientation_covariance
        wc = msg.angular_velocity_covariance
        ac = msg.linear_acceleration_covariance

        self.get_logger().info(
            (
                f"stamp={t_sec:.6f}s  frame={msg.header.frame_id}  recv_dt={dt_ms:.3f}ms\n"
                f"  orientation (x,y,z,w): [{q.x:.6f}, {q.y:.6f}, {q.z:.6f}, {q.w:.6f}]\n"
                f"  angular_velocity (rad/s): [{w.x:.6f}, {w.y:.6f}, {w.z:.6f}]\n"
                f"  linear_accel (m/s^2): [{a.x:.6f}, {a.y:.6f}, {a.z:.6f}]\n"
                f"  cov_diag(ori, ang, acc): "
                f"[{oc[0]:.6f},{oc[4]:.6f},{oc[8]:.6f} | "
                f"{wc[0]:.6f},{wc[4]:.6f},{wc[8]:.6f} | "
                f"{ac[0]:.6f},{ac[4]:.6f},{ac[8]:.6f}]"
            )
        )


def main():
    rclpy.init()
    node = ImuPrinterPy()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
